src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -843,6 +859,23 @@ public SolutionState WithRunAnalyzers(ProjectId projectId, bool runAnalyzers)
             return ForkProject(newProject);
         }
 
+        /// <summary>
+        /// Create a new solution instance with the project specified updated to have
+        /// the specified attributes, except for <see cref="ProjectInfo.ProjectAttributes.Version"/> which is auto-incremented.
+        /// </summary>
+        public SolutionState WithProjectAttributes(ProjectId projectId, ProjectInfo.ProjectAttributes attributes)
+        {
+            var oldProject = GetRequiredProjectState(projectId);
+            var newProject = oldProject.WithAttributes(attributes);
+
+            if (oldProject == newProject)
+            {
+                return this;
+            }
+
+            return ForkProject(newProject);
```

Comment:
This is even trickier: the project attributes changing could mean you have to change the document states too, but it's less clear here.